### ISSUE TITLE
protoc-gen-go: emit versioning information

### ISSUE THREAD

**Issue by dsnet** [golang/protobuf#524]
The generator currently emits the following:
```
// Code generated by protoc-gen-go. DO NOT EDIT.
// source: extension_user/extension_user.proto
```

We should improve this by adding information about the versions:
1) What `go` toolchain was used (e.g., `go1.9.6`)
2) What version of `protoc` was used (e.g., `libprotoc 3.5.1`)
3) What version of `protoc-gen-go` was used

The 3rd item is harder to do, but the upcoming version control features in the Go toolchain will help with this. My understanding is that one of the features it will provide is a cryptographic signing of all the Go packages that went into building that binary.

For now, I would be happy to just see items 1 and 2 start being encoded in the generated infolog.

**Comment by dsnet** [comment]
Furthermore, any special options passed to protoc-gen-go should probably be provided in that log. For example, whether the `grpc` plugin is used.

**Comment by dsnet** [comment]
Related: #384

**Comment by neild** [comment]
If we do this, we should also make "make regenerate" produce an error when run with the "wrong" (non-canonical) version of the toolchain.

**Comment by dsnet** [comment]
> If we do this, we should also make "make regenerate" produce an error when run with the "wrong" (non-canonical) version of the toolchain.

That sounds fine to me. However, I don't think that's this issue, but what @tamird is trying address in #521.

**Comment by dsnet** [comment]
If it wasn't clear, this issue should address both generated code and also provide a command-line flag to check this information.

**Comment by dsnet** [comment]
This is blocked on https://github.com/golang/go/issues/26404

**Comment by zeeraw** [comment]
What is the status on the command line flag?

**Comment by dsnet** [comment]
Still blocked on golang/go#26404, which is slated for release in Go1.12.

**Comment by metasyn** [comment]
Sounds like it is unblocked now, with go 1.12 released & golang/go#24604 closed?

**Comment by rocco-haro** [comment]
Any updates on this?

**Comment by saravana-thiyagaraj** [comment]
Any updates on this?

**Comment by thsaravana** [comment]
@dsnet Any update on this?

**Comment by RahulMitra** [comment]
@dsnet  any update on this?

**Comment by dsnet** [comment]
This is already resolved in the v2 branch. The v2 release is targeting end of Q1.

**Comment by dsnet** [comment]
This is fixed in the [v1.20.0 release](https://github.com/protocolbuffers/protobuf-go/releases/tag/v1.20.0) of `protoc-gen-go`.
